File: sdks/javascript/src/embedding.ts
/**
 * Simple embedding/vector operations
 */

import * as crypto from 'crypto';

/**
 * Create a simple embedding vector from text
 * Uses hash-based approach for deterministic results
 */
export function embed(text: string, dimensions: number = 64): number[] {
  // Create hash
  const hash = crypto.createHash('sha256').update(text).digest();

  // Convert to vector
  const vec: number[] = [];
  for (let i = 0; i < dimensions; i++) {
    const idx = i % hash.length;
    const val = hash[idx] / 255; // Normalize to 0-1
    vec.push(val - 0.5); // Center around 0
  }

  // Normalize to unit vector
  const norm = Math.sqrt(vec.reduce((sum, v) => sum + v * v, 0));
  if (norm > 0) {
    return vec.map((v) => v / norm);
  }

  return vec;
}

/**
 * Calculate cosine similarity between two vectors
 */
export function cosineSimilarity(a: number[], b: number[]): number {
  if (a.length !== b.length) return 0;

  let dot = 0;
  let normA = 0;
  let normB = 0;

  for (let i = 0; i < a.length; i++) {
    dot += a[i] * b[i];
    normA += a[i] * a[i];
    normB += b[i] * b[i];
  }

  if (normA === 0 || normB === 0) return 0;

  return dot / (Math.sqrt(normA) * Math.sqrt(normB));
}

/**
 * Calculate resonance match
 */
export function resonanceMatch(focus: number[], signalResonance: number[]): number {
  return cosineSimilarity(focus, signalResonance);
}

File: sdks/python/agenthighway/embedding.py
"""
Simple embedding/vector operations

In production, use sentence-transformers or OpenAI embeddings.
This is a lightweight fallback.
"""

import hashlib
import numpy as np
from typing import List


def embed(text: str, dimensions: int = 64) -> List[float]:
    """
    Create a simple embedding vector from text.
    
    Uses hash-based approach for deterministic results.
    For production, replace with proper embeddings.
    
    Args:
        text: Input text
        dimensions: Vector dimensions (default 64)
        
    Returns:
        Normalized vector
    """
    # Create hash
    hash_obj = hashlib.sha256(text.encode())
    hash_bytes = hash_obj.digest()
    
    # Convert to vector
    vec = []
    for i in range(dimensions):
        # Use different parts of hash
        idx = i % len(hash_bytes)
        val = hash_bytes[idx] / 255.0  # Normalize to 0-1
        vec.append(val - 0.5)  # Center around 0
    
    # Normalize to unit vector
    norm = np.linalg.norm(vec)
    if norm > 0:
        vec = [v / norm for v in vec]
    
    return vec


def cosine_similarity(a: List[float], b: List[float]) -> float:
    """Calculate cosine similarity between two vectors."""
    if len(a) != len(b):
        return 0.0
    
    dot = sum(x * y for x, y in zip(a, b))
    norm_a = sum(x * x for x in a) ** 0.5
    norm_b = sum(x * x for x in b) ** 0.5
    
    if norm_a == 0 or norm_b == 0:
        return 0.0
    
    return dot / (norm_a * norm_b)


def resonance_match(focus: List[float], signal_resonance: List[float]) -> float:
    """
    Calculate how well a signal resonates with an agent's focus.
    
    Returns value 0-1 where 1 is perfect match.
    """
    return cosine_similarity(focus, signal_resonance)
